pymc_extras/inference/laplace_approx/laplace.py

```diff
@@ -147,189 +139,84 @@ def get_conditional_gaussian_approximation(
-    frozen_model = freeze_dims_and_data(model)
-
-    if not transform_samples:
-        untransformed_model = remove_value_transforms(frozen_model)
-        logp = untransformed_model.logp(jacobian=False)
-        variables = untransformed_model.continuous_value_vars
-    else:
-        logp = frozen_model.logp(jacobian=True)
-        variables = frozen_model.continuous_value_vars
-
-    variable_names = {var.name for var in variables}
-    optimized_free_params = {k: v for k, v in optimized_point.items() if k in variable_names}
-    mu = DictToArrayBijection.map(optimized_free_params)
-
-    f_fused, _ = scipy_optimize_funcs_from_loss(
-        loss=-logp,
-        inputs=variables,
-        initial_point_dict=optimized_free_params,
-        use_grad=True,
-        use_hess=True,
-        use_hessp=False,
-        gradient_backend=gradient_backend,
-        compile_kwargs=compile_kwargs,
-    )
-
-    H = -f_fused(mu.data)[-1]
-    if H.ndim == 1:
-        H = np.expand_dims(H, axis=1)
-    H_inv = np.linalg.pinv(np.where(np.abs(H) < zero_tol, 0, -H))
-
-    def stabilize(x, jitter):
-        return x + np.eye(x.shape[0]) * jitter
-
-    H_inv = H_inv if diag_jitter is None else stabilize(H_inv, diag_jitter)
+def model_to_laplace_approx(
+    model: pm.Model, unpacked_variable_names: list[str], chains: int = 1, draws: int = 500
+):
+    # temp_chain and temp_draw are a hack to allow sampling from the Laplace approximation. We only have one mu and cov,
+    # so we add batch dims (which correspond to chains and draws). But the names "chain" and "draw" are reserved.
+    coords = model.coords | {
+        "temp_chain": np.arange(chains),
+        "temp_draw": np.arange(draws),
+        "unpacked_variable_names": unpacked_variable_names,
+    }
+    free_rvs = model.free_RVs
+
+    initial_point = model.initial_point()
+    raveled_vars = DictToArrayBijection.map(initial_point)
+    raveled_shape = raveled_vars.data.shape[0]
+
+    with pm.Model(coords=coords, model=None) as laplace_model:
+        mu = pm.Flat("mean_vector", shape=(raveled_shape,))
+        cov = pm.Flat("covariance_matrix", shape=(raveled_shape, raveled_shape))
+        laplace_approximation = pm.MvNormal(
+            "laplace_approximation",
+            mu=mu,
+            cov=cov,
+            dims=["temp_chain", "temp_draw", "unpacked_variable_names"],
+        )
 
-    try:
-        np.linalg.cholesky(H_inv)
-    except np.linalg.LinAlgError:
-        if on_bad_cov == "error":
-            raise np.linalg.LinAlgError(
-                "Inverse Hessian not positive-semi definite at the provided point"
+        cursor = 0
+        for rv, info in zip(free_rvs, raveled_vars.point_map_info):
+            transform = model.rvs_to_transforms[rv]
+            name, shape, size, dtype = info
+            unconstrained_value = (
+                laplace_approximation[..., cursor : cursor + size]
+                .reshape((chains, draws, *shape))
```

Comment:
Can you add a test with a variable that has different dimensionality in the unconstrained space (like a dirichlet)?

Reply:
Nevermind figured it out